File: tests/r_parity/81_didm.py
"""StatsPAI dCDH 2020 DID_M parity (Python side) -- Module 81.

Runs ``sp.did_multiplegt`` on a non-absorbing panel where treatment switches
both on and off; the companion R script runs ``DIDmultiplegt::did_multiplegt``
on the same dumped CSV.

Version note that makes this module possible at all: the CRAN package's 2.x
rewrite routes the classic estimator through ``mode="old"``, and that path
returns ``NaN`` even on the package's own bundled example. The archived
**0.1.4** is the last release where it works, and it is what this pins
against. See the R script for the install line.

Emitted: the static DID_M effect, the dynamic effect at horizon 1, and the
placebo at lag 1 -- the three the reference reports for this call. The
placebo is requested with ``placebo_sign="r"``: dCDH's own Stata and R
implementations disagree on its sign, and this module compares against R. Standard
errors are not compared: the R side runs ``brep=0`` and StatsPAI's come from
its own bootstrap.

Tolerance: rel < 1e-6.
"""

from __future__ import annotations

import warnings

import numpy as np
import pandas as pd

from _common import ParityRecord, dump_csv, write_results

MODULE = "81_didm"


def _panel() -> pd.DataFrame:
    """Treatment switches on AND off -- the design DID_M exists for.

    An absorbing panel would leave the switch-off branch and its sign
    convention untested, which is where this estimator differs from every
    cohort-based one.
    """
    rng = np.random.default_rng(9)
    rows = []
    for uid in range(1, 181):
        fe = rng.normal()
        d_prev = 0
        for t in range(1, 7):
            if t == 1:
                d = int(rng.random() < 0.3)
            elif rng.random() < 0.25:
                d = 1 - d_prev
            else:
                d = d_prev
            d_prev = d
            rows.append(
                {
                    "id": uid,
                    "t": t,
                    "d": d,
                    "y": fe + 0.25 * t + 1.2 * d + rng.normal(0, 0.6),
                }
            )
    return pd.DataFrame(rows)


def main() -> None:
    import statspai as sp

    df = _panel()
    dump_csv(df, MODULE)

    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        res = sp.did_multiplegt(
            df,
            y="y",
            group="id",
            time="t",
            treatment="d",
            placebo=1,
            dynamic=1,
            n_boot=0,
            # dCDH's Stata and R implementations disagree on the placebo's
            # SIGN -- on did::mpdta both give |0.024269| and the three
            # effects agree to six decimals, but Stata reports it positive
            # and DIDmultiplegt negative. This module compares against R,
            # so it asks for R's convention; the default stays on Stata's.
            placebo_sign="r",
        )
    es = res.model_info["event_study"].set_index("relative_time")["att"]

    rows = [
        ParityRecord(
            module=MODULE,
            side="py",
            statistic="effect",
            estimate=float(es.loc[0]),
            n=int(len(df)),
        ),
        ParityRecord(
            module=MODULE,
            side="py",
            statistic="dynamic_1",
            estimate=float(es.loc[1]),
            n=int(len(df)),
        ),
        ParityRecord(
            module=MODULE,
            side="py",
            statistic="placebo_1",
            estimate=float(es.loc[-1]),
            n=int(len(df)),
        ),
    ]
    write_results(MODULE, "py", rows, extra={"estimator": "did_multiplegt"})


if __name__ == "__main__":
    main()

File: tests/r_parity/81_didm.R
# StatsPAI dCDH 2020 DID_M parity (R side) -- Module 81.
#
# Runs DIDmultiplegt::did_multiplegt on the panel dumped by 81_didm.py.
#
# ⚠️ Version matters. The CRAN package's 2.x rewrite routes the classic
# estimator through mode="old", and that path returns NaN even on the
# package's own bundled `wagepan_mgt` example. The archived 0.1.4 is the
# last release where it works:
#
#   install.packages("assertthat")
#   install.packages(
#     "https://cran.r-project.org/src/contrib/Archive/DIDmultiplegt/DIDmultiplegt_0.1.4.tar.gz",
#     repos = NULL, type = "source", lib = <a private library>)
#
# Point it at that library with STATSPAI_DIDM_LIB, otherwise this falls back
# to the default library and will fail loudly rather than emit 2.x's NaN.
#
# Standard errors are not emitted: this runs brep = 0.
#
# Tolerance: rel < 1e-6.

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) {
  dirname(normalizePath(sub("^--file=", "", .file_arg[1])))
} else {
  getwd()
}
source(file.path(.script_dir, "_common.R"))

.didm_lib <- Sys.getenv("STATSPAI_DIDM_LIB")
if (nzchar(.didm_lib)) .libPaths(c(.didm_lib, .libPaths()))
suppressPackageStartupMessages(library(DIDmultiplegt))

.ver <- as.character(utils::packageVersion("DIDmultiplegt"))
if (utils::compareVersion(.ver, "2.0.0") >= 0) {
  stop(sprintf(
    paste0("DIDmultiplegt %s is installed; its mode='old' path returns NaN ",
           "even on the package's own example. Install 0.1.4 into a private ",
           "library and point STATSPAI_DIDM_LIB at it."), .ver))
}

MODULE <- "81_didm"
df <- as.data.frame(read_csv_strict(MODULE))

res <- suppressWarnings(
  did_multiplegt(df = df, Y = "y", G = "id", T = "t", D = "d",
                 placebo = 1, dynamic = 1, brep = 0)
)

rows <- list()
emit <- function(nm, value) {
  rows[[length(rows) + 1]] <<- parity_row(
    module = MODULE, statistic = nm, estimate = as.numeric(value),
    se = NA_real_, n = nrow(df)
  )
}
emit("effect", res$effect)
emit("dynamic_1", res$dynamic_1)
emit("placebo_1", res$placebo_1)

write_results(MODULE, rows, extra = list(
  reference = "DIDmultiplegt::did_multiplegt (archived 0.1.4)",
  DIDmultiplegt = .ver
))
